### DISCUSSION TITLE
add support prisma-json-types-generator

### DISCUSSION THREAD

**Discussion by davidkukert** [m1212e/prismabox#9] [Ideas]
Hi, I have a suggestion, add the types from prisma-json-types-generator to the schema generated for typebox.

**Comment by m1212e** [comment]
Hey, thanks for your contribution! Could you elaborate a bit more? What behavior are you suggesting exactly? Maybe even with a "this would work in an ideal world" code example on how you would like to use prismabox in conjunction with prisma-json-types-generator? Thanks!

**Reply by davidkukert** [reply]
prisma-json-types-generator adds custom types for json columns, the idea would be to take this type if possible and build a schema to use based on it.

**Reply by davidkukert** [reply]
Example:
schema.prisma
```
model Role {
    id          String   @id @db.VarChar(26)
    name        String
    slug        String   @unique
    /// [RolePermissions]
    permissions Json
    createdAt   DateTime @default(now()) @map("created_at")
    updatedAt   DateTime @default(now()) @updatedAt @map("updated_at")

    @@map("tb_roles")
}
```

index.d.ts
```
import type { Prisma } from '@prisma/client'

type RecordOptionalKey<K extends string | number | symbol, T> = { [P in K]?: T }

type PermissionsAction =
	| 'readList'
	| 'readDetails'
	| 'create'
	| 'update'
	| 'delete'

type PermissionsEntityModel = Prisma.ModelName | '*'

declare global {
	namespace PrismaJson {
		type RolePermissions = RecordOptionalKey<
			PermissionsAction,
			PermissionsEntityModel[]
		>
	}
}

```

and based on the code above generate the following schema for example:

`
```
import type {
	PermissionsAction,
	PermissionsEntityModel,
} from '@/types/permissions'
import { t } from 'elysia'

export const permissions = t.Record(
	t.Union([
		t.Literal<PermissionsAction>('readList'),
		t.Literal<PermissionsAction>('readDetails'),
		t.Literal<PermissionsAction>('create'),
		t.Literal<PermissionsAction>('update'),
		t.Literal<PermissionsAction>('delete'),
	]),
	t.Optional(
		t.Array(
			t.Union([
				t.Literal<PermissionsEntityModel>('Chapter'),
				t.Literal<PermissionsEntityModel>('Project'),
				t.Literal<PermissionsEntityModel>('Role'),
				t.Literal<PermissionsEntityModel>('Tag'),
				t.Literal<PermissionsEntityModel>('User'),
				t.Literal<PermissionsEntityModel>('*'),
			]),
		),
	),
)
```
`

**Reply by m1212e** [reply]
As far as I see inside the prisma schema there is no information available on which structure the assigned type has? This would add the additional step of parsing the Typescript code and add a schema based on the tsc output which is wayyy out of scope for this project!

**Reply by m1212e** [reply]
Do you know of some project which does these kinds of things? Converting TS into a typebox schema? If this exists somewhere this would be way easier to achieve. Without I unfortunately don't see any realistic way for prismabox to actually implement your suggestion :(

**Reply by davidkukert** [reply]
I understand, I thought it would be possible.

**Comment by davidkukert** [comment]
so instead of adding support for prisma-json-types-generator, add the ability to put custom typebox schemas. Where instead of the lib creating the schema for a given column, the dev can place his own already created schema. And prismabox only uses it instead of t.Any for example in the JSON col..

**Comment by davidkukert** [comment]
based on the idea of ​​prisma-json-types-generator to add types to json fields and use in primebox, where the dev can get his own schema and use it instead of the t.Any used in JSON fields currently by prismabox.

**Reply by m1212e** [reply]
I see, but at this point, what is preventing you from just defining the schema as a TS variable and using composite/other utils to compose a matching schema for your final schema? E.g. if you have some fields which typebox outputs as `any` you can use `Type.Omit` to remove these fields from the schema and then `Type.Compose` to join the cleaned schema with your custom ones. I think this would result in a more readable approach since you lose the ability for multiline and actually linted/parsed code when defining the schema in a prisma comment. Whats your take on this, in which scenarios would you see this as beneficial?

**Reply by davidkukert** [reply]
you did not understand. After finishing building the database structure and using Prismabox to generate the schemas, I then make changes to the fields that need to be changed. But since you don't want to add new features to Prismabox, that's fine. It was just an idea.

**Reply by m1212e** [reply]
Sure, thanks for your contribution. Always happy to review a PR!
